reconcile.go

```diff
@@ -32,18 +32,30 @@ func (a *Autopilot) reconcile() error {
 	// have the promoter calculate the required Raft changeset.
 	changes := a.promoter.CalculatePromotionsAndDemotions(conf, state)
 
+	// Apply the demotions to the failed servers first, but only once per a
+	// reconciliation round and if the number of voters is odd. In that case we
+	// don't want to start with the promotions, since that temporalily inflates
+	// quorum, which could lead to cluster failure. If we did apply any, then stop
+	// here as we do not want to apply the promotions at the same time as a means of
+	// preventing cluster instability.
+	if len(state.Voters)%2 != 0 {
+		if done, err := a.applyDemotions(state, changes, true); done {
+			return err
+		}
+	}
+
 	// apply the promotions, if we did apply any then stop here
 	// as we do not want to apply the demotions at the same time
 	// as a means of preventing cluster instability.
 	if done, err := a.applyPromotions(state, changes); done {
 		return err
 	}
 
-	// apply the demotions, if we did apply any then stop here
+	// apply the demotions to healthy servers, if we did apply any then stop here
```

Comment:
🤔 Hmmm. What happens to unhealthy servers when they don't get handled above now? For example if there is an even number of voters we won't demote _any_ unhealthy servers at all until we've promoted another and got back to an odd number of voters. It's not immediately obvious to me if that is the correct behaviour although it might be - even number of voters is not usually a target state.

Even if we have odd voters though, we will only process one unhealthy demotion per reconcile round. I _guess_ this is OK in the sense that if that works, next reconcile round we can demote one more etc...

Reply:
Sorry, this comment took a bit longer than the others because I needed to think about it and I ended up testing it empirically to make sure I didn't misunderstand the logic. 

TLDR is that it should be fine, since AP promotes, then demotes over two reconcile rounds, taking care of the unhealthy voter. Here is a step-by-step that I wrote up for myself, but I think is worth sharing.

**t0 (starting point):** 
- 4 redundancy zones
- 8 nodes
- 4 voters
- quorum of 3

| RZ1 | RZ2 | RZ3 | RZ4 |
| --- | --- | --- | --- |
| A*  | B*  | C*  | D*  |
| E   | F   | G   | H   |

**t1: `D` becomes unhealthy**

| RZ1 | RZ2 | RZ3 | RZ4    |
| --- | --- | --- | ------ |
| A*  | B*  | C*  | ~~D*~~ |
| E   | F   | G   | H      |

**t2 (reconcile loop 1):** 
- AP skips demoting `D`, because we still have an even number of voters (four).
- AP promotes `H`, we have 5 voters, quorum is still 3.

| RZ1 | RZ2 | RZ3 | RZ4    |
| --- | --- | --- | ------ |
| A*  | B*  | C*  | ~~D*~~ |
| E   | F   | G   | H*     |

**t3 (reconcile loop 2):**
- AP demotes `D` because it's unhealthy and we have an odd number of voters going into the reconcile round (five).
- We now have 4 voters, quorum was stable the whole time at 3.

| RZ1 | RZ2 | RZ3 | RZ4   |
| --- | --- | --- | ----- |
| A*  | B*  | C*  | ~~D~~ |
| E   | F   | G   | H*    |

One thing I am noticing here, is that we could possibly demote an unhealthy voter after the promotions in the same loop to avoid waiting the extra 10 seconds 🤔 I would welcome any opinions here. 

> Even if we have odd voters though, we will only process one unhealthy demotion per reconcile round. I guess this is OK in the sense that if that works, next reconcile round we can demote one more etc...

That is true. I guess we could look into demoting more than one at a time, but then we would need to add some logic to guard the `min_quorum` value, which currently (as in with the proposed changes) is guaranteed to be maintained because we only demote one voter and only when we have an odd number of them. On one hand, this seems like more complexity, on the other maybe it's more explicit. I am not sure if that optimisation is worth complicating the code more.